File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {

    List<String> ans;

    // Condition, max open & close = n
    public void create(String curr, int open, int close, int n) {
        if (open == n && close == n) ans.add(curr);
        if (open < n) create(curr + "(", open + 1, close, n);
        if(close < open) create(curr + ")", open, close + 1, n);
    }

    public List<String> generateParenthesis(int n) {
        ans = new ArrayList<>();
        create("(", 1, 0, n);
        return ans;
    }
}

File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:

    void create(string curr, int open, int close, int n, vector<string>& ans) {
        if(open == n && close == n) ans.push_back(curr);
        if(open < n) create(curr + "(", open + 1, close, n, ans);
        if(close < open) create(curr + ")", open, close + 1, n, ans);
    }

    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        create("(", 1, 0, n, ans);
        return ans;
    }
};